mlir/lib/Conversion/TosaToMLProgram/TosaToMLProgramPass.cpp

```diff
@@ -0,0 +1,50 @@
+//===- TosaToMLProgramPass.cpp - Lowering Tosa to MLProgram Dialect
+//-------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transformation pass legalizes the TOSA dialect to the MLProgram dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/TosaToMLProgram/TosaToMLProgram.h"
+#include "mlir/Dialect/MLProgram/IR/MLProgram.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Tosa/Transforms/Passes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_TOSATOMLPROGRAM
+#include "mlir/Conversion/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+using namespace tosa;
+
+namespace {
+struct TosaToMLProgram : public impl::TosaToMLProgramBase<TosaToMLProgram> {
+public:
+  void runOnOperation() override {
+    auto *context = &getContext();
+    auto moduleOp = getOperation();
+
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+    target.addLegalDialect<ml_program::MLProgramDialect>();
+    target.addIllegalOp<tosa::VariableOp, tosa::VariableReadOp,
+                        tosa::VariableWriteOp>();
+    target.markUnknownOpDynamicallyLegal([](Operation *) { return true; });
+
+    mlir::tosa::populateTosaToMLProgramConversionPatterns(&patterns);
+
+    if (failed(applyPartialConversion(moduleOp, target, std::move(patterns))))
+      signalPassFailure();
```

Comment:
You should log an error via `emitMatchFailure`

Reply:
sorry, I couldn't find anything about `emitMatchFailure` [[Link](https://www.google.com/search?sca_esv=575897283&q=mlir+%22emitMatchFailure%22&sa=X&ved=2ahUKEwjN79Kxk42CAxXoMjQIHXWiAt4Q5t4CegQIExAB&biw=1844&bih=866&dpr=2)]. When I tried to build it, it also fails. Other passes are also using `signalPassFailure()`

Reply:
I actually don't think this one can fail ... I think it would only fail if this didn't converge after 10 iterations given partial conversion. But fine as is.

(probably meant notifyMatchFailure)

Reply:
actually copied from TosaToLinag here: https://github.com/llvm/llvm-project/blob/d199fd76f7b76d902d8ef210d82689f299934793/mlir/lib/Conversion/TosaToLinalg/TosaToLinalgNamedPass.cpp#L70C3-L70C3. 

And all other TosaTo* passes are checking this if condition and do `signalPassFailure` 
If `notifyMatchFailure` is the correct one to use here, I suggest I can do another patch to update all the TosaTo* passes?